File: typescript/src/ancilis/episodes/signed.ts
/** Native authentication under caller-provisioned trust; never reconstruction. */
import { createHash, createPrivateKey, createPublicKey, KeyObject, sign, verify } from "node:crypto";
import { types } from "node:util";
import { z } from "zod";
import { canonicalEpisodeJSON, detached, hashEpisodePayload, nativeIdSchema, nativeTimeSchema, now } from "./contract.js";
import type { ContentReference, EpisodeSnapshot } from "./contract.js";
import { verifyEpisodeSnapshot } from "./verification.js";
export const MAX_SIGNED_EXPORT_BYTES = 32 * 1024 * 1024;
const DOMAIN = "ancilis-signed-episode/1\n";
const hex = (size: number) => z.string().regex(new RegExp(`^[0-9a-f]{${size}}(?![\\s\\S])`));
const keySchema = z.object({ key_id: nativeIdSchema, public_key: hex(64), not_before: nativeTimeSchema.nullable(), not_after: nativeTimeSchema.nullable(), revoked: z.boolean() }).strict();
const policySchema = z.object({
    schema: z.literal("ancilis-episode-trust-policy/1"), tenant: nativeIdSchema, source: nativeIdSchema,
    keys: z.array(keySchema).max(64), body_mode: z.enum(["NONE", "ALL_REFERENCED"]),
    max_body_bytes: z.number().int().min(1).max(16777216), max_total_body_bytes: z.number().int().min(1).max(67108864), max_body_requests: z.number().int().min(1).max(1024),
}).strict();
export type EpisodeTrustKey = z.infer<typeof keySchema>;
export type EpisodeTrustPolicyDocument = z.infer<typeof policySchema>;
export class SignedEpisodeError extends Error {
    constructor(readonly code: string) { super(code); this.name = "SignedEpisodeError"; }
}
function freeze<T>(value: T): T {
    if (value !== null && typeof value === "object") {
        for (const child of Object.values(value))
            freeze(child);
        Object.freeze(value);
    }
    return value;
}
const arrayPrototype = Object.getPrototypeOf(Uint8Array.prototype);
const intrinsicLength = Object.getOwnPropertyDescriptor(arrayPrototype, "byteLength")!.get!;
const intrinsicBuffer = Object.getOwnPropertyDescriptor(arrayPrototype, "buffer")!.get!;
const intrinsicOffset = Object.getOwnPropertyDescriptor(arrayPrototype, "byteOffset")!.get!;
function byteLength(value: Uint8Array): number { return Reflect.apply(intrinsicLength, value, []) as number; }
function copyBytes(value: Uint8Array): Buffer {
    const length = byteLength(value);
    const buffer = Reflect.apply(intrinsicBuffer, value, []) as ArrayBuffer;
    const offset = Reflect.apply(intrinsicOffset, value, []) as number;
    return Buffer.from(new Uint8Array(buffer, offset, length));
}
function scalars(text: string): void {
    for (let i = 0; i < text.length; i++) {
        const code = text.charCodeAt(i);
        if (code >= 0xd800 && code <= 0xdbff) {
            const next = text.charCodeAt(++i);
            if (!(next >= 0xdc00 && next <= 0xdfff))
                throw Error();
        }
        else if (code >= 0xdc00 && code <= 0xdfff)
            throw Error();
    }
}
/** No key discovery: this policy must come from an independently trusted channel. */
const policyKeys = new WeakMap<EpisodeTrustPolicy, ReadonlyMap<string, KeyObject>>();
const signerKeys = new WeakMap<EpisodeSigner, KeyObject>();

export class EpisodeTrustPolicy {
    readonly sha256: string;
    readonly #document: EpisodeTrustPolicyDocument;
    constructor(document: EpisodeTrustPolicyDocument) {
        try {
            const value = policySchema.parse(detached(document));
            const importedKeys = new Map<string, KeyObject>();
            let previous = "";
            for (const key of value.keys) {
                if (key.key_id <= previous || (key.not_before !== null && key.not_after !== null && key.not_before >= key.not_after))
                    throw Error();
                const bytes = Buffer.from(key.public_key, "hex");
                const imported = createPublicKey({ key: Buffer.concat([Buffer.from("302a300506032b6570032100", "hex"), bytes]), format: "der", type: "spki" });
                if (imported.asymmetricKeyType !== "ed25519" || Buffer.from(imported.export({ format: "jwk" }).x!, "base64url").toString("hex") !== key.public_key)
                    throw Error();
                importedKeys.set(key.key_id, imported);
                previous = key.key_id;
            }
            policyKeys.set(this, importedKeys);
            this.#document = freeze(value);
            this.sha256 = hashEpisodePayload("ancilis-episode-trust-policy/1", value);
            Object.freeze(this);
        }
        catch {
            throw new SignedEpisodeError("INVALID_TRUST_POLICY");
        }
    }
    toJSON(): EpisodeTrustPolicyDocument { return detached(this.#document); }

}
export interface EpisodeSignerOptions {
    tenant: string;
    source: string;
    keyId: string;
    privateKey: KeyObject;
}
export class EpisodeSigner {
    readonly tenant: string;
    readonly source: string;
    readonly keyId: string;
    constructor(options: EpisodeSignerOptions) {
        try {
            this.tenant = nativeIdSchema.parse(options.tenant);
            this.source = nativeIdSchema.parse(options.source);
            this.keyId = nativeIdSchema.parse(options.keyId);
            if (!(options.privateKey instanceof KeyObject) || options.privateKey.type !== "private" || options.privateKey.asymmetricKeyType !== "ed25519")
                throw Error();
            signerKeys.set(this, options.privateKey);
            Object.freeze(this);
        }
        catch {
            throw new SignedEpisodeError("INVALID_SIGNER");
        }
    }
    static fromSeed(options: Omit<EpisodeSignerOptions, "privateKey"> & {
        seed: Uint8Array;
    }): EpisodeSigner {
        try {
            if (!(options.seed instanceof Uint8Array) || !types.isUint8Array(options.seed) || byteLength(options.seed) !== 32)
                throw Error();
            const seed = copyBytes(options.seed);
            const key = createPrivateKey({ key: Buffer.concat([Buffer.from("302e020100300506032b657004220420", "hex"), seed]), format: "der", type: "pkcs8" });
            if (!Buffer.from(key.export({ format: "jwk" }).d!, "base64url").equals(seed))
                throw Error();
            return new EpisodeSigner({ ...options, privateKey: key });
        }
        catch {
            throw new SignedEpisodeError("INVALID_SIGNER");
        }
    }
}
export function signEpisodeSnapshot(snapshot: EpisodeSnapshot, signer: EpisodeSigner): string {
    const privateKey = signerKeys.get(signer);
    if (!(signer instanceof EpisodeSigner) || !privateKey)
        throw new SignedEpisodeError("INVALID_SIGNER");
    let value: EpisodeSnapshot;
    try {
        value = detached(snapshot);
    }
    catch {
        throw new SignedEpisodeError("INVALID_NATIVE_SNAPSHOT");
    }
    const reason = nativeReason(value, now());
    if (reason)
        throw new SignedEpisodeError(reason);
    if (value.tenant !== signer.tenant || value.open.owner_source !== signer.source)
        throw new SignedEpisodeError("SCOPE_MISMATCH");
    const unsigned = { schema: "ancilis-signed-episode/1", algorithm: "Ed25519", key_id: signer.keyId, snapshot: value };
    const canonical = canonicalEpisodeJSON(unsigned);
    if (Buffer.byteLength(canonical) + 143 > MAX_SIGNED_EXPORT_BYTES)
        throw new SignedEpisodeError("EXPORT_TOO_LARGE");
    let signature: string;
    try {
        signature = sign(null, Buffer.from(DOMAIN + canonical), privateKey).toString("hex");
    }
    catch {
        throw new SignedEpisodeError("SIGNATURE_VERIFICATION_ERROR");
    }
    return canonicalEpisodeJSON({ ...unsigned, signature });
}

function nativeReason(snapshot: unknown, assessedAt: string): string | null {
    const checked = verifyEpisodeSnapshot(snapshot, { assessedAt });
    if (checked.status === "REJECTED")
        return "INVALID_NATIVE_SNAPSHOT";
    return checked.reasons.includes("NATIVE_HISTORY_DISCARDED") ? "NATIVE_HISTORY_DISCARDED" : null;
}
export interface ProtectedBodyRequest {
    readonly tenant: string;
    readonly episode: string;
    readonly open_sha256: string;
    readonly revision_id: string;
    readonly event_id: string;
    readonly reference: Readonly<Omit<ContentReference, "classification_receipt_refs"> & {
        classification_receipt_refs: readonly string[];
    }>;
}
export type BodyResolver = (request: ProtectedBodyRequest, signal?: AbortSignal) => Uint8Array | null | Promise<Uint8Array | null>;
export interface SignedVerificationOptions {
    trust: EpisodeTrustPolicy;
    assessedAt?: string;
    bodyResolver?: BodyResolver;
    signal?: AbortSignal;
}
export interface SignedEpisodeVerification {
    readonly schema: "ancilis-signed-verification/1";
    readonly status: "AUTHENTICATED" | "UNVERIFIED" | "REJECTED" | "ERROR";
    readonly envelope_authenticated: boolean;
    readonly protected_bodies: "NOT_REQUESTED" | "VERIFIED" | "NO_REFERENCES" | "UNAVAILABLE" | "MISMATCH" | "ERROR" | "LIMIT_EXCEEDED";
    readonly reconstruction: "UNSUPPORTED";
    readonly policy_sha256: string;
    readonly assessed_at: string;
    readonly export_sha256: string | null;
    readonly verified_body_count: number;
    readonly reasons: readonly string[];
    readonly verified_claim_refs: [
    ];
}
type Result = { -readonly [K in keyof SignedEpisodeVerification]: SignedEpisodeVerification[K] };
function out(result: Result, status: Result["status"], reason: string, bodies: Result["protected_bodies"] = "NOT_REQUESTED"): Result {
    result.status = status;
    result.protected_bodies = bodies;
    result.reasons = [...(result.envelope_authenticated ? ["ENVELOPE_AUTHENTICATED"] : []), reason];
    return result;
}
const envelopeSchema = z.object({ schema: z.literal("ancilis-signed-episode/1"), algorithm: z.literal("Ed25519"), key_id: nativeIdSchema, signature: hex(128), snapshot: z.unknown().refine(value => value !== undefined) }).strict();
function prepare(wire: string | Uint8Array, options: SignedVerificationOptions): [
    Result,
    ProtectedBodyRequest[]
] {
    if (!(options.trust instanceof EpisodeTrustPolicy))
        throw new SignedEpisodeError("INVALID_TRUST_POLICY");
    const assessed = options.assessedAt ?? now();
    if (!nativeTimeSchema.safeParse(assessed).success)
        throw new SignedEpisodeError("INVALID_ASSESSMENT_TIME");
    const result: Result = { schema: "ancilis-signed-verification/1", status: "UNVERIFIED", envelope_authenticated: false, protected_bodies: "NOT_REQUESTED", reconstruction: "UNSUPPORTED", policy_sha256: options.trust.sha256, assessed_at: assessed, export_sha256: null, verified_body_count: 0, reasons: [], verified_claim_refs: [] };
    let raw: Buffer, value: unknown;
    try {
        if (typeof wire === "string") {
            if (wire.length > MAX_SIGNED_EXPORT_BYTES)
                return [out(result, "REJECTED", "EXPORT_TOO_LARGE"), []];
            scalars(wire);
            raw = Buffer.from(wire, "utf8");
        }
        else if (wire instanceof Uint8Array && types.isUint8Array(wire)) {
            if (byteLength(wire) > MAX_SIGNED_EXPORT_BYTES)
                return [out(result, "REJECTED", "EXPORT_TOO_LARGE"), []];
            raw = copyBytes(wire);
        }
        else
            throw Error();
        if (raw.length > MAX_SIGNED_EXPORT_BYTES)
            return [out(result, "REJECTED", "EXPORT_TOO_LARGE"), []];
        value = JSON.parse(new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(raw));
        if (!Buffer.from(canonicalEpisodeJSON(value)).equals(raw))
            throw Error();
    }
    catch {
        return [out(result, "REJECTED", "INVALID_SIGNED_EXPORT"), []];
    }
    result.export_sha256 = createHash("sha256").update(raw).digest("hex");
    const parsed = envelopeSchema.safeParse(value);
    if (!parsed.success)
        return [out(result, "REJECTED", "INVALID_SIGNED_EXPORT"), []];
    const envelope = parsed.data;
    const snapshot = envelope.snapshot as EpisodeSnapshot;
    const reason = nativeReason(snapshot, assessed);
    if (reason)
        return [out(result, "REJECTED", reason), []];
    const policy = options.trust.toJSON();
    if (snapshot.tenant !== policy.tenant || snapshot.open.owner_source !== policy.source)
        return [out(result, "REJECTED", "SCOPE_MISMATCH"), []];
    const key = policy.keys.find(k => k.key_id === envelope.key_id);
    if (!key)
        return [out(result, "UNVERIFIED", "UNTRUSTED_KEY"), []];
    if (key.revoked)
        return [out(result, "UNVERIFIED", "REVOKED_KEY"), []];
    if ((key.not_before !== null && assessed < key.not_before) || (key.not_after !== null && assessed >= key.not_after))
        return [out(result, "UNVERIFIED", "KEY_NOT_CURRENT"), []];
    const { signature, ...unsigned } = envelope;
    try {
        if (!verify(null, Buffer.from(DOMAIN + canonicalEpisodeJSON(unsigned)), policyKeys.get(options.trust)!.get(key.key_id)!, Buffer.from(signature, "hex")))
            return [out(result, "REJECTED", "INVALID_SIGNATURE"), []];
    }
    catch {
        return [out(result, "ERROR", "SIGNATURE_VERIFICATION_ERROR"), []];
    }
    result.envelope_authenticated = true;
    if (policy.body_mode === "NONE")
        return [out(result, "AUTHENTICATED", "BODIES_NOT_REQUESTED"), []];
    const requests: ProtectedBodyRequest[] = [];
    let total = 0;
    for (const row of snapshot.observations)
        for (const reference of row.artifacts) {
            total += reference.byte_length;
            if (requests.length >= policy.max_body_requests || reference.byte_length > policy.max_body_bytes || total > policy.max_total_body_bytes)
                return [out(result, "UNVERIFIED", "BODY_LIMIT_EXCEEDED", "LIMIT_EXCEEDED"), []];
            requests.push(freeze({ tenant: snapshot.tenant, episode: snapshot.episode, open_sha256: snapshot.open_sha256, revision_id: snapshot.revision_id, event_id: row.event_id, reference: detached(reference) }));
        }
    if (!requests.length)
        out(result, "UNVERIFIED", "NO_BODY_REFERENCES", "NO_REFERENCES");
    return [result, requests];
}
/** Resolve nothing before authentication. Cancellation remains a caller action. */
export async function verifySignedEpisode(wire: string | Uint8Array, options: SignedVerificationOptions): Promise<SignedEpisodeVerification> {
    if (options.bodyResolver !== undefined && typeof options.bodyResolver !== "function")
        throw new SignedEpisodeError("INVALID_BODY_RESOLVER");
    options.signal?.throwIfAborted();
    const [result, requests] = prepare(wire, options);
    for (const request of requests) {
        options.signal?.throwIfAborted();
        let body: unknown;
        try {
            body = options.bodyResolver ? await options.bodyResolver(request, options.signal) : null;
        }
        catch {
            options.signal?.throwIfAborted();
            return freeze(out(result, "ERROR", "BODY_RESOLVER_ERROR", "ERROR"));
        }
        options.signal?.throwIfAborted();
        if (body === null)
            return freeze(out(result, "UNVERIFIED", "BODY_UNAVAILABLE", "UNAVAILABLE"));
        if (!(body instanceof Uint8Array) || !types.isUint8Array(body))
            return freeze(out(result, "ERROR", "BODY_RESOLVER_ERROR", "ERROR"));
        try {
            if (byteLength(body) !== request.reference.byte_length)
                return freeze(out(result, "REJECTED", "BODY_MISMATCH", "MISMATCH"));
            const copy = copyBytes(body);
            if (copy.length !== request.reference.byte_length || createHash("sha256").update(copy).digest("hex") !== request.reference.sha256)
                return freeze(out(result, "REJECTED", "BODY_MISMATCH", "MISMATCH"));
        }
        catch {
            return freeze(out(result, "ERROR", "BODY_RESOLVER_ERROR", "ERROR"));
        }
        result.verified_body_count++;
    }
    if (requests.length)
        out(result, "AUTHENTICATED", "BODIES_VERIFIED", "VERIFIED");
    return freeze(result);
}

File: python/src/ancilis/episodes/signed.py
"""Scoped native authentication and explicit protected-byte verification.

Signatures authenticate collector assertions, never classification or completeness.
Cryptography is loaded only when signing/trust configuration is constructed.
"""

from __future__ import annotations

import dataclasses
import hashlib
import inspect
import json
import re
from collections.abc import Awaitable, Callable, Mapping
from types import MappingProxyType
from typing import TYPE_CHECKING, Any, Literal, TypedDict, cast

from . import (
    ContentEvidence,
    EpisodeSnapshot,
    EpisodeSnapshotDict,
    _freeze,
    _hash,
    _id,
    _now,
    _thaw,
    _timestamp,
    canonical_json,
    verify_episode_snapshot,
)

if TYPE_CHECKING:
    from cryptography.exceptions import InvalidSignature
    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
        Ed25519PrivateKey,
        Ed25519PublicKey,
    )

MAX_SIGNED_EXPORT_BYTES = 32 * 1024 * 1024
_DOMAIN = b"ancilis-signed-episode/1\n"


class SignedEpisodeError(ValueError):
    """Fixed-code configuration/signing error; never contains evidence or keys."""


def _crypto() -> tuple[type[Ed25519PrivateKey], type[Ed25519PublicKey], type[InvalidSignature]]:
    try:
        from cryptography.exceptions import InvalidSignature
        from cryptography.hazmat.primitives.asymmetric.ed25519 import (
            Ed25519PrivateKey,
            Ed25519PublicKey,
        )
    except ImportError:
        raise SignedEpisodeError("CRYPTO_UNAVAILABLE") from None
    return Ed25519PrivateKey, Ed25519PublicKey, InvalidSignature


def _hex(value: object, size: int) -> bool:
    return (
        isinstance(value, str) and re.fullmatch(r"[0-9a-f]{" + str(size) + "}", value) is not None
    )


class EpisodeTrustKey(TypedDict):
    key_id: str
    public_key: str
    not_before: str | None
    not_after: str | None
    revoked: bool


class EpisodeTrustPolicyDict(TypedDict):
    schema: Literal["ancilis-episode-trust-policy/1"]
    tenant: str
    source: str
    keys: list[EpisodeTrustKey]
    body_mode: Literal["NONE", "ALL_REFERENCED"]
    max_body_bytes: int
    max_total_body_bytes: int
    max_body_requests: int


@dataclasses.dataclass(frozen=True, init=False)
class EpisodeTrustPolicy:
    """Independently provisioned, immutable key admission and body requirements."""

    _document: object = dataclasses.field(repr=False)
    _keys: Mapping[str, Ed25519PublicKey] = dataclasses.field(repr=False)
    sha256: str

    def __init__(self, document: Mapping[str, Any]) -> None:
        _, public_type, _ = _crypto()
        try:
            d = json.loads(canonical_json(document))
            if set(d) != set(EpisodeTrustPolicyDict.__annotations__):
                raise ValueError()
            if d["schema"] != "ancilis-episode-trust-policy/1" or d["body_mode"] not in (
                "NONE",
                "ALL_REFERENCED",
            ):
                raise ValueError()
            _id(d["tenant"])
            _id(d["source"])
            for name, maximum in (
                ("max_body_bytes", 16777216),
                ("max_total_body_bytes", 67108864),
                ("max_body_requests", 1024),
            ):
                if type(d[name]) is not int or not 1 <= d[name] <= maximum:
                    raise ValueError()
            keys = d["keys"]
            if type(keys) is not list or len(keys) > 64:
                raise ValueError()
            imported = {}
            previous = ""
            for key in keys:
                if type(key) is not dict or set(key) != set(EpisodeTrustKey.__annotations__):
                    raise ValueError()
                key_id = _id(key["key_id"])
                if (
                    key_id <= previous
                    or not _hex(key["public_key"], 64)
                    or type(key["revoked"]) is not bool
                ):
                    raise ValueError()
                for time in (key["not_before"], key["not_after"]):
                    if time is not None:
                        _timestamp(time)
                if (
                    key["not_before"] is not None
                    and key["not_after"] is not None
                    and key["not_before"] >= key["not_after"]
                ):
                    raise ValueError()
                imported[key_id] = public_type.from_public_bytes(bytes.fromhex(key["public_key"]))
                previous = key_id
            object.__setattr__(self, "_document", _freeze(d))
            object.__setattr__(self, "_keys", MappingProxyType(imported))
            object.__setattr__(self, "sha256", _hash("ancilis-episode-trust-policy/1", d))
        except Exception:
            raise SignedEpisodeError("INVALID_TRUST_POLICY") from None

    def to_dict(self) -> EpisodeTrustPolicyDict:
        return cast(EpisodeTrustPolicyDict, _thaw(self._document))


@dataclasses.dataclass(frozen=True)
class EpisodeSigner:
    tenant: str
    source: str
    key_id: str
    private_key: Ed25519PrivateKey = dataclasses.field(repr=False)

    def __post_init__(self) -> None:
        private_type, _, _ = _crypto()
        try:
            _id(self.tenant)
            _id(self.source)
            _id(self.key_id)
            if not isinstance(self.private_key, private_type):
                raise ValueError()
        except Exception:
            raise SignedEpisodeError("INVALID_SIGNER") from None

    @classmethod
    def from_seed(cls, tenant: str, source: str, key_id: str, *, seed: bytes) -> EpisodeSigner:
        private_type, _, _ = _crypto()
        try:
            if type(seed) is not bytes or len(seed) != 32:
                raise ValueError()
            key = private_type.from_private_bytes(bytes(seed))
        except Exception:
            raise SignedEpisodeError("INVALID_SIGNER") from None
        return cls(tenant, source, key_id, key)


def _native_reason(snapshot: object, assessed: str) -> str | None:
    checked = verify_episode_snapshot(
        cast(Mapping[str, Any], snapshot), assessed_at=assessed
    ).to_dict()
    if checked["status"] == "REJECTED":
        return "INVALID_NATIVE_SNAPSHOT"
    if "NATIVE_HISTORY_DISCARDED" in checked["reasons"]:
        return "NATIVE_HISTORY_DISCARDED"
    return None


def sign_episode_snapshot(
    snapshot: EpisodeSnapshot | Mapping[str, Any], signer: EpisodeSigner
) -> str:
    if not isinstance(signer, EpisodeSigner):
        raise SignedEpisodeError("INVALID_SIGNER")
    try:
        value = json.loads(
            canonical_json(
                snapshot.to_dict() if isinstance(snapshot, EpisodeSnapshot) else snapshot
            )
        )
        reason = _native_reason(value, _now())
    except Exception:
        raise SignedEpisodeError("INVALID_NATIVE_SNAPSHOT") from None
    if reason:
        raise SignedEpisodeError(reason)
    if value["tenant"] != signer.tenant or value["open"]["owner_source"] != signer.source:
        raise SignedEpisodeError("SCOPE_MISMATCH")
    unsigned = {
        "schema": "ancilis-signed-episode/1",
        "algorithm": "Ed25519",
        "key_id": signer.key_id,
        "snapshot": value,
    }
    encoded = canonical_json(unsigned)
    # The signature member adds 143 canonical bytes, including its separator.
    if len(encoded) + 143 > MAX_SIGNED_EXPORT_BYTES:
        raise SignedEpisodeError("EXPORT_TOO_LARGE")
    try:
        signature = signer.private_key.sign(_DOMAIN + encoded).hex()
    except Exception:
        raise SignedEpisodeError("SIGNATURE_VERIFICATION_ERROR") from None
    return canonical_json({**unsigned, "signature": signature}).decode("utf-8")


SignedStatus = Literal["AUTHENTICATED", "UNVERIFIED", "REJECTED", "ERROR"]
BodyStatus = Literal[
    "NOT_REQUESTED",
    "VERIFIED",
    "NO_REFERENCES",
    "UNAVAILABLE",
    "MISMATCH",
    "ERROR",
    "LIMIT_EXCEEDED",
]


class SignedEpisodeVerificationDict(TypedDict):
    schema: Literal["ancilis-signed-verification/1"]
    status: SignedStatus
    envelope_authenticated: bool
    protected_bodies: BodyStatus
    reconstruction: Literal["UNSUPPORTED"]
    policy_sha256: str
    assessed_at: str
    export_sha256: str | None
    verified_body_count: int
    reasons: list[str]
    verified_claim_refs: list[str]


@dataclasses.dataclass(frozen=True)
class SignedEpisodeVerification:
    _document: object = dataclasses.field(repr=False)

    def to_dict(self) -> SignedEpisodeVerificationDict:
        return cast(SignedEpisodeVerificationDict, _thaw(self._document))


@dataclasses.dataclass(frozen=True)
class ProtectedBodyRequest:
    tenant: str
    episode: str
    open_sha256: str
    revision_id: str
    event_id: str
    reference: ContentEvidence


_BodyPlan = tuple[ProtectedBodyRequest, int, str]


BodyResolver = Callable[[ProtectedBodyRequest], bytes | None]
AsyncBodyResolver = Callable[[ProtectedBodyRequest], bytes | None | Awaitable[bytes | None]]


def _out(
    result: SignedEpisodeVerificationDict,
    status: SignedStatus,
    reason: str,
    bodies: BodyStatus = "NOT_REQUESTED",
) -> None:
    result["status"] = status
    result["protected_bodies"] = bodies
    result["reasons"] = (["ENVELOPE_AUTHENTICATED"] if result["envelope_authenticated"] else []) + [
        reason
    ]


def _prepare(
    export: str | bytes, trust: EpisodeTrustPolicy, assessed_at: str | None
) -> tuple[SignedEpisodeVerificationDict, list[_BodyPlan]]:
    if not isinstance(trust, EpisodeTrustPolicy):
        raise SignedEpisodeError("INVALID_TRUST_POLICY")
    assessed = assessed_at if assessed_at is not None else _now()
    try:
        _timestamp(assessed)
    except Exception:
        raise SignedEpisodeError("INVALID_ASSESSMENT_TIME") from None
    result: SignedEpisodeVerificationDict = {
        "schema": "ancilis-signed-verification/1",
        "status": "UNVERIFIED",
        "envelope_authenticated": False,
        "protected_bodies": "NOT_REQUESTED",
        "reconstruction": "UNSUPPORTED",
        "policy_sha256": trust.sha256,
        "assessed_at": assessed,
        "export_sha256": None,
        "verified_body_count": 0,
        "reasons": [],
        "verified_claim_refs": [],
    }
    try:
        if type(export) not in (str, bytes):
            raise ValueError()
        if len(export) > MAX_SIGNED_EXPORT_BYTES:
            _out(result, "REJECTED", "EXPORT_TOO_LARGE")
            return result, []
        raw = export.encode("utf-8", errors="strict") if isinstance(export, str) else export
        if len(raw) > MAX_SIGNED_EXPORT_BYTES:
            _out(result, "REJECTED", "EXPORT_TOO_LARGE")
            return result, []
        value = json.loads(raw.decode("utf-8", errors="strict"))
        if canonical_json(value) != raw:
            raise ValueError()
    except Exception:
        _out(result, "REJECTED", "INVALID_SIGNED_EXPORT")
        return result, []
    result["export_sha256"] = hashlib.sha256(raw).hexdigest()
    try:
        if type(value) is not dict or set(value) != {
            "schema",
            "algorithm",
            "key_id",
            "snapshot",
            "signature",
        }:
            raise ValueError()
        if (
            value["schema"] != "ancilis-signed-episode/1"
            or value["algorithm"] != "Ed25519"
            or not _hex(value["signature"], 128)
        ):
            raise ValueError()
        _id(value["key_id"])
    except Exception:
        _out(result, "REJECTED", "INVALID_SIGNED_EXPORT")
        return result, []
    snapshot = cast(EpisodeSnapshotDict, value["snapshot"])
    reason = _native_reason(snapshot, assessed)
    if reason:
        _out(result, "REJECTED", reason)
        return result, []
    policy = trust.to_dict()
    if (
        snapshot["tenant"] != policy["tenant"]
        or snapshot["open"]["owner_source"] != policy["source"]
    ):
        _out(result, "REJECTED", "SCOPE_MISMATCH")
        return result, []
    key = next((k for k in policy["keys"] if k["key_id"] == value["key_id"]), None)
    if key is None:
        _out(result, "UNVERIFIED", "UNTRUSTED_KEY")
        return result, []
    if key["revoked"]:
        _out(result, "UNVERIFIED", "REVOKED_KEY")
        return result, []
    if (key["not_before"] is not None and assessed < key["not_before"]) or (
        key["not_after"] is not None and assessed >= key["not_after"]
    ):
        _out(result, "UNVERIFIED", "KEY_NOT_CURRENT")
        return result, []
    _, _, invalid_signature = _crypto()
    unsigned = {k: v for k, v in value.items() if k != "signature"}
    try:
        trust._keys[key["key_id"]].verify(
            bytes.fromhex(value["signature"]), _DOMAIN + canonical_json(unsigned)
        )
    except invalid_signature:
        _out(result, "REJECTED", "INVALID_SIGNATURE")
        return result, []
    except Exception:
        _out(result, "ERROR", "SIGNATURE_VERIFICATION_ERROR")
        return result, []
    result["envelope_authenticated"] = True
    if policy["body_mode"] == "NONE":
        _out(result, "AUTHENTICATED", "BODIES_NOT_REQUESTED")
        return result, []
    requests: list[_BodyPlan] = []
    total = 0
    for row in snapshot["observations"]:
        for reference in row["artifacts"]:
            total += reference["byte_length"]
            if (
                len(requests) >= policy["max_body_requests"]
                or reference["byte_length"] > policy["max_body_bytes"]
                or total > policy["max_total_body_bytes"]
            ):
                _out(result, "UNVERIFIED", "BODY_LIMIT_EXCEEDED", "LIMIT_EXCEEDED")
                return result, []
            frozen = ContentEvidence(
                reference["artifact"],
                reference["sha256"],
                reference["byte_length"],
                reference["role"],
                reference["access_scope"],
                tuple(reference["classification_receipt_refs"]),
            )
            request = ProtectedBodyRequest(
                snapshot["tenant"],
                snapshot["episode"],
                snapshot["open_sha256"],
                snapshot["revision_id"],
                row["event_id"],
                frozen,
            )
            # Callback-visible dataclasses are advisory frozen objects. Keep the
            # cryptographic expectations separately, never read them back from
            # a callback argument after user code runs.
            requests.append((request, reference["byte_length"], reference["sha256"]))
    if not requests:
        _out(result, "UNVERIFIED", "NO_BODY_REFERENCES", "NO_REFERENCES")
    return result, requests


def _check_body(
    result: SignedEpisodeVerificationDict,
    expected_length: int,
    expected_sha256: str,
    body: object,
) -> bool:
    if body is None:
        _out(result, "UNVERIFIED", "BODY_UNAVAILABLE", "UNAVAILABLE")
        return False
    if type(body) is not bytes:
        if inspect.iscoroutine(body):
            body.close()
        _out(result, "ERROR", "BODY_RESOLVER_ERROR", "ERROR")
        return False
    if len(body) != expected_length or hashlib.sha256(body).hexdigest() != expected_sha256:
        _out(result, "REJECTED", "BODY_MISMATCH", "MISMATCH")
        return False
    result["verified_body_count"] += 1
    return True


def _resolver_config(resolver: object, *, asynchronous: bool) -> None:
    if resolver is not None and (
        not callable(resolver)
        or (
            not asynchronous
            and (
                inspect.iscoroutinefunction(resolver)
                or inspect.iscoroutinefunction(type(resolver).__call__)
            )
        )
    ):
        raise SignedEpisodeError("INVALID_BODY_RESOLVER")


def verify_signed_episode(
    export: str | bytes,
    trust: EpisodeTrustPolicy,
    *,
    assessed_at: str | None = None,
    body_resolver: BodyResolver | None = None,
) -> SignedEpisodeVerification:
    _resolver_config(body_resolver, asynchronous=False)
    result, requests = _prepare(export, trust, assessed_at)
    for request, expected_length, expected_sha256 in requests:
        try:
            body = body_resolver(request) if body_resolver else None
        except Exception:
            _out(result, "ERROR", "BODY_RESOLVER_ERROR", "ERROR")
            break
        if not _check_body(result, expected_length, expected_sha256, body):
            break
    else:
        if requests:
            _out(result, "AUTHENTICATED", "BODIES_VERIFIED", "VERIFIED")
    return SignedEpisodeVerification(_freeze(result))


async def averify_signed_episode(
    export: str | bytes,
    trust: EpisodeTrustPolicy,
    *,
    assessed_at: str | None = None,
    body_resolver: AsyncBodyResolver | None = None,
) -> SignedEpisodeVerification:
    _resolver_config(body_resolver, asynchronous=True)
    result, requests = _prepare(export, trust, assessed_at)
    for request, expected_length, expected_sha256 in requests:
        try:
            body = body_resolver(request) if body_resolver else None
            if inspect.isawaitable(body):
                body = await body
        except Exception:
            _out(result, "ERROR", "BODY_RESOLVER_ERROR", "ERROR")
            break
        if not _check_body(result, expected_length, expected_sha256, body):
            break
    else:
        if requests:
            _out(result, "AUTHENTICATED", "BODIES_VERIFIED", "VERIFIED")
    return SignedEpisodeVerification(_freeze(result))
